test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -230,25 +299,232 @@ func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, clust
 	}
 
-	dockerMachinePool.Status.Ready = len(dockerMachinePool.Spec.ProviderIDList) == int(*machinePool.Spec.Replicas)
+	if len(dockerMachinePool.Spec.ProviderIDList) == int(*machinePool.Spec.Replicas) {
+		dockerMachinePool.Status.Ready = true
+		conditions.MarkTrue(dockerMachinePool, expv1.ReplicasReadyCondition)
+
+		return ctrl.Result{}, nil
+	}
+
+	dockerMachinePool.Status.Ready = false
+	conditions.MarkFalse(dockerMachinePool, expv1.ReplicasReadyCondition, expv1.WaitingForReplicasReadyReason, clusterv1.ConditionSeverityInfo, "")
 
 	// if some machine is still provisioning, force reconcile in few seconds to check again infrastructure.
-	if !dockerMachinePool.Status.Ready && res.IsZero() {
+	if res.IsZero() {
 		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
 	}
+
 	return res, nil
 }
 
+func getDockerMachines(ctx context.Context, c client.Client, cluster clusterv1.Cluster, machinePool expv1.MachinePool, dockerMachinePool infraexpv1.DockerMachinePool) (*infrav1.DockerMachineList, error) {
+	dockerMachineList := &infrav1.DockerMachineList{}
+	labels := map[string]string{
+		clusterv1.ClusterNameLabel:     cluster.Name,
+		clusterv1.MachinePoolNameLabel: machinePool.Name,
+	}
+	if err := c.List(ctx, dockerMachineList, client.InNamespace(dockerMachinePool.Namespace), client.MatchingLabels(labels)); err != nil {
+		return nil, err
+	}
+
+	return dockerMachineList, nil
+}
+
+// DeleteOrphanedDockerMachines deletes any DockerMachines owned by the DockerMachinePool that reference an invalid providerID, i.e. not in the latest copy of the node pool instances.
+func (r *DockerMachinePoolReconciler) DeleteOrphanedDockerMachines(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool, nodePoolMachineStatuses []docker.NodePoolMachineStatus) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.V(2).Info("Deleting orphaned machines", "dockerMachinePool", dockerMachinePool.Name, "namespace", dockerMachinePool.Namespace, "nodePoolMachineStatuses", nodePoolMachineStatuses)
+	dockerMachineList, err := getDockerMachines(ctx, r.Client, *cluster, *machinePool, *dockerMachinePool)
+	if err != nil {
+		return err
+	}
+
+	log.V(2).Info("DockerMachineList kind is", "kind", dockerMachineList.GetObjectKind())
+
+	nodePoolMachineNames := map[string]struct{}{}
+	for _, nodePoolMachineStatus := range nodePoolMachineStatuses {
+		nodePoolMachineNames[nodePoolMachineStatus.Name] = struct{}{}
+	}
+
+	for i := range dockerMachineList.Items {
+		dockerMachine := &dockerMachineList.Items[i]
+		if _, ok := nodePoolMachineNames[dockerMachine.Name]; !ok {
+			machine, err := util.GetOwnerMachine(ctx, r.Client, dockerMachine.ObjectMeta)
+			// Not found doesn't return an error, so we need to check for nil.
+			if err != nil {
+				return errors.Wrapf(err, "error getting owner Machine for DockerMachine %s/%s", dockerMachine.Namespace, dockerMachine.Name)
+			}
+			if machine == nil {
```

Comment:
Will we ever reach this condition when machine will be nil? If k8s cant get hold of the machine, it will return an error. And if it finds, it will return a value. I think probably we should check for errors.IsNotFound(err) and ignore if the error is of that type

Reply:
Based on the [implementation](https://github.com/kubernetes-sigs/cluster-api/blob/54dfd423d80e42f79c33c0149f9719dba3863f11/util/util.go#L242) of `GetOwnerMachine()`, it returns `nil, nil` if the ownerRefs do not have a Machine type. It returns an error if we failed the API call or to parse the GVK. So it seems like an `errors.IsNotFound(err)` would only be applicable if the owning Machine was deleted.

I think in general, we could try to be more consistent about documenting the expected behavior of functions that return `obj, err` when `err == nil`. I generally expect it to return `obj, nil` or `nil, err`, meaning if there's no error, we can expect a result.